amethyst_animation/src/skinning/systems.rs

```diff
@@ -17,19 +17,27 @@ use super::resources::*;
 /// System for performing vertex skinning.
 ///
 /// Needs to run after global transforms have been updated for the current frame.
-#[derive(Default, Debug)]
+#[derive(Debug)]
 pub struct VertexSkinningSystem {
     /// Also scratch space, used while determining which skins need to be updated.
     updated: BitSet,
     updated_skins: BitSet,
     /// Used for tracking modifications to global transforms
-    updated_id: Option<ReaderId<ComponentEvent>>,
+    updated_id: ReaderId<ComponentEvent>,
 }
 
 impl VertexSkinningSystem {
     /// Creates a new `VertexSkinningSystem`
-    pub fn new() -> Self {
-        Self::default()
+    pub fn new(world: &mut World) -> Self {
+        use amethyst_core::ecs::prelude::SystemData;
+        <Self as System<'_>>::SystemData::setup(&mut world.res);
```

Comment:
Since the `setup` trait method is no longer overridden, this line can be omitted from the `System` (ditto for all other systems that no longer override `setup`).

**edit:** ohh, the system data needs to set up the world for things like `EventChannel` to exist. Wonder if there's a nice way to avoid needing every one to add this line at the start of `new`. Perhaps:

```rust
trait System {
    fn new(world: &mut World) -> Self;
    // ..
}

// Perhaps the Dispatcher would construct the `System`:
dispatcher.add_system::<VertexSkinningSystem>("skinning_system", &[]);

// Then inside `add_system`:
fn add_system<S: System>(&mut self, name: &str, deps: &[&str]) -> Result<_, _> {
    S::setup(&mut self.world);
    let system = S::new(&mut self.world);
    // ..
}
```

It doesn't work so nicely when there are systems that take parameters in constructors like this:

```rust
FreeRotationSystem::new(world, self.sensitivity_x, self.sensitivity_y),
```

 (using an associated type for the "rest of the parameters" is ugly 🙁)

@torkleyy do you have a preference?
I'm happy to live with calling `SystemData::setup` for now, and we can revisit this after.

Reply:
I believe this may be resolved for now.

Reply:
oh yes certainly